lib/src/rules/cascade_invocations.dart

```diff
@@ -266,12 +266,41 @@ class _CascadableExpression {
       !_hasCriticalDependencies(expressionBox);
 
   bool _hasCriticalDependencies(_CascadableExpression expressionBox) {
-    bool isCriticalNode(AstNode node) =>
-        node.canonicalElement == expressionBox.element;
-    return expressionBox.isCritical &&
-        criticalNodes.any((node) =>
-            isCriticalNode(node) ||
-            node.traverseNodesInDFS().any(isCriticalNode));
+    if (!expressionBox.isCritical) return false;
+
+    for (var node in criticalNodes) {
+      if (_NodeVisitor(expressionBox).isOrHasCriticalNode(node)) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+}
+
+class _NodeVisitor extends GeneralizingAstVisitor {
+  final _CascadableExpression expressionBox;
+
+  bool foundCriticalNode = false;
+  _NodeVisitor(this.expressionBox);
+
+  bool isCriticalNode(AstNode node) =>
+      node.canonicalElement == expressionBox.element;
+
+  bool isOrHasCriticalNode(AstNode node) {
+    if (isCriticalNode(node)) return true;
```

Comment:
As per a comment on a different PR, when we visit the `node` we'll repeat this test. Unless we gain a performance benefit from skipping the visit, we should remove this line to avoid duplication.